### PR TITLE
Convert directory_parser and factory to pytest

### PR DESCRIPTION
**Pull request by rgpt**
* Converted directory_parser and factory to pytest
* Checked with Pylint.

### CODE REVIEW

**Comment by rgpt** [comment]
@whimboo PR Added. Please Review.
I believe Coverage decrease 0.1 % due to stop using of deprecated function that is rmtree in the last commit https://github.com/mozilla/mozdownload/pull/537/commits/408e854c7253c139c7f3775cb3d80aa0c5f9afad#diff-86a953f2df0c74587c942c584cae8800R40

File: tests/directory_parser/test_directory_parser.py [hunk 1]
```diff
@@ -9,59 +9,56 @@
...
+
+def test_init(httpd):
+    """Testing the basic functionality of the DirectoryParser Class"""
+
+    # DirectoryParser returns output
+    parser = DirectoryParser(httpd.get_url())
+
+    # relies on the presence of other files in the directory
+    # Checks if DirectoryParser lists the server entries
+    assert parser.entries != [], "parser.entries were not listed"
```
Comment (whimboo) [inline tests/directory_parser/test_directory_parser.py]: Does the message actually hide the real values if the assertion fails? I cannot remember how pytest handles that. If yes, we should remove the message.

File: tests/directory_parser/test_directory_parser.py [hunk 2]
```diff
@@ -9,59 +9,56 @@
...
+
+    # path_regex to mozdownload -t release -p win32 -v latest
+    testpath = urljoin(httpd.get_url(), 'directoryparser/')
+    parser1 = DirectoryParser(testpath)
```
Comment (whimboo) [inline tests/directory_parser/test_directory_parser.py]: Just `parser` please.

File: tests/directory_parser/test_directory_parser.py [hunk 3]
```diff
@@ -9,59 +9,56 @@
...
+    parser1.entries.sort()
+    testdir = os.listdir(urljoin(httpd.router.doc_root, 'directoryparser'))
+    testdir.sort()
+    assert parser1.entries == testdir
+
+
+def test_filter(httpd):
+    """Testing the DirectoryParser filter method"""
+
+    parser = DirectoryParser(urljoin(httpd.get_url(), 'directoryparser', 'filter/'))
```
Comment (whimboo) [inline tests/directory_parser/test_directory_parser.py]: If the next non-empty line after a docstring is not a comment, don't add an empty line in between.

File: tests/directory_parser/test_directory_parser.py [hunk 4]
```diff
@@ -9,59 +9,56 @@
...
+
+    # Get the contents of the folder - dirs and files
+    folder_path = urljoin(httpd.router.doc_root, 'directoryparser', 'filter')
+    contents = os.listdir(folder_path)
+    contents.sort()
+    assert parser.entries == contents
```
Comment (whimboo) [inline tests/directory_parser/test_directory_parser.py]: Interesting. Why don't we have to sort the parser entries here?

File: tests/directory_parser/test_directory_parser.py [hunk 5]
```diff
@@ -9,59 +9,56 @@
...
+    assert parser.entries == dirs
+
+    # Test filter method with a function
+    parser.entries = parser.filter(lambda x: x == dirs[0])
+    assert parser.entries == [dirs[0]]
+
+
+def test_names_with_spaces(httpd):
+
+    parser = DirectoryParser(urljoin(httpd.get_url(), 'directoryparser', 'some spaces/'))
```
Comment (whimboo) [inline tests/directory_parser/test_directory_parser.py]: No empty line above please.

File: tests/directory_parser/test_directory_parser.py [hunk 6]
```diff
@@ -9,59 +9,56 @@
...
+
+    # Get the contents of the folder - dirs and files
+    folder_path = urljoin(httpd.router.doc_root, 'directoryparser', 'some spaces')
+    contents = os.listdir(folder_path)
+    contents.sort()
+    assert parser.entries == contents
```
Comment (whimboo) [inline tests/directory_parser/test_directory_parser.py]: Same here as above for sorting parser entries.

File: tests/factory/test_factory_invalid_options.py [hunk 7]
```diff
@@ -2,186 +2,151 @@
...
+    """Test that missing mandatory options for release builds raise an exception"""
+    with pytest.raises(ValueError):
+        FactoryScraper('release', destination=str(tmpdir), base_url=httpd.get_url())
 
-tests = [
+
+def test_candidate_without_version(httpd, tmpdir):
+    """Test that missing mandatory options for candidate builds raise an exception"""
+    with pytest.raises(ValueError):
+        FactoryScraper('release', destination=str(tmpdir), base_url=httpd.get_url())
```
Comment (whimboo) [inline tests/factory/test_factory_invalid_options.py]: There is a bug in the test. This should be `candidate` and not `release`.

File: tests/factory/test_factory_invalid_options.py [hunk 8]
```diff
@@ -2,186 +2,151 @@
...
+
+
+def test_try_without_revision(httpd, tmpdir):
+    """Test that missing mandatory options for try builds raise an exception"""
+    with pytest.raises(ValueError):
+        FactoryScraper('try', destination=str(tmpdir), base_url=httpd.get_url())
+
+
+def test_non_daily_fennec(httpd, tmpdir):
+    """Test that non-daily scrapper_type for fennec raises exception"""
```
Comment (whimboo) [inline tests/factory/test_factory_invalid_options.py]: Please fix the typing failure: `scraper`. It also applies to the next test.

File: tests/factory/test_factory_invalid_options.py [hunk 9]
```diff
@@ -2,186 +2,151 @@
...
-        self.assertRaises(NotSupportedError, FactoryScraper,
-                          scraper_type='daily',
-                          destination=self.temp_dir,
-                          base_url=self.wdir,
-                          logger=self.logger,
-                          application='devedition',
-                          version='60.0b1')
-
-
-class TestFactoryUnusedOptions(mhttpd.MozHttpdBaseTest):
```
Comment (whimboo) [inline tests/factory/test_factory_invalid_options.py]: Please lets move all the tests from this class to it's own file, and name both files similar to the class names.

File: tests/directory_parser/test_directory_parser.py (hunk 1 above)
Comment (rgpt) [reply tests/directory_parser/test_directory_parser.py]: It do not hide real values and just adds more information which is the message. So no problem here by using message.

File: tests/directory_parser/test_directory_parser.py (hunk 4 above)
Comment (rgpt) [reply tests/directory_parser/test_directory_parser.py]: Actually Data in both parser.entries and contents are already sorted and so sort() function not doing anything at all over the data. Hence, in these functions without sorting operation was working.
I am removing sort function as it is a redundant operation and there is no use in these test cases.

File: tests/directory_parser/test_directory_parser.py (hunk 6 above)
Comment (rgpt) [reply tests/directory_parser/test_directory_parser.py]: referenced in the above comments.

**Comment by rgpt** [comment]
@whimboo PR updated. Please Review.

File: tests/directory_parser/test_directory_parser.py (hunk 1 above)
Comment (whimboo) [reply tests/directory_parser/test_directory_parser.py]: Good, so at least here we have an improvement compared to the Python unittest assertions.

File: tests/directory_parser/test_directory_parser.py (hunk 4 above)
Comment (whimboo) [reply tests/directory_parser/test_directory_parser.py]: Yes, that is what I was wondering about. So lets also sort the parser entries.

File: tests/does_not_exist.html [hunk 10]
```diff
@@ -0,0 +1 @@
+temp
```
Comment (whimboo) [inline tests/does_not_exist.html]: What is the reason for that file?

File: tests/factory/test_factory_invalid_options.py [hunk 11]
```diff
@@ -1,187 +1,42 @@
-import copy
```
Comment (whimboo) [inline tests/factory/test_factory_invalid_options.py]: The file name should contain `missing_mandatory_options` to mimic the former class name.

File: tests/factory/test_factory_scraper.py [hunk 12]
```diff
@@ -2,88 +2,62 @@
...
 import mozfile
-from mock import patch
+import pytest
 
 from mozdownload import FactoryScraper
-import mozhttpd_base_test as mhttpd
 
 
-tests = [
+@pytest.mark.parametrize('scraper_type,builds,fname,arg', [
```
Comment (whimboo) [inline tests/factory/test_factory_scraper.py]: please use `filename,args`.

File: tests/factory/test_factory_scraper.py [hunk 13]
```diff
@@ -2,88 +2,62 @@
...
+    kwargs = copy.deepcopy(base_kwargs)
+    kwargs.update(arg)
+
+    build = FactoryScraper(scraper_type, **kwargs)
+    build.download()
+
+    dir_content = os.listdir(str(tmpdir))
+    assert fname in dir_content
+
+    mozfile.remove(os.path.join(str(tmpdir), fname))
```
Comment (whimboo) [inline tests/factory/test_factory_scraper.py]: The file is in the temporary directory and as such will automatically be removed when the `tmpdir` fixture runs its clean-up steps. Also we want to get rid of the `mozfile` dependency.

File: tests/factory/test_factory_unused_options.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+import os
+
+import mozfile
+import pytest
+
+from mozdownload.utils import urljoin
+from mozdownload import FactoryScraper
+
+
+@pytest.mark.parametrize('scraper_type,builds,fname,arg', [
```
Comment (whimboo) [inline tests/factory/test_factory_unused_options.py]: Same as for the other file with `filename,args`.

File: tests/factory/test_factory_unused_options.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    kwargs = copy.deepcopy(base_kwargs)
+    kwargs.update(arg)
+
+    build = FactoryScraper(scraper_type, **kwargs)
+    build.download()
+
+    dir_content = os.listdir(str(tmpdir))
+    assert fname in dir_content
+
+    mozfile.remove(os.path.join(str(tmpdir), fname))
```
Comment (whimboo) [inline tests/factory/test_factory_unused_options.py]: Lets get rid of this line. It's not needed as explained above.

File: tests/factory/test_factory_unused_options.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+from mozdownload.utils import urljoin
+from mozdownload import FactoryScraper
+
+
+@pytest.mark.parametrize('scraper_type,builds,fname,arg', [
+    # ReleaseScraper
+    ('release', None, 'firefox-23.0.1.en-US.win32.exe', {
+        'version': '23.0.1',
+        'platform': 'win32',
+
```
Comment (whimboo) [inline tests/factory/test_factory_unused_options.py]: Maybe you can add a comment line `# unused options` instead of an empty line? It would help to better understand why those args are separated here.

File: tests/mozhttpd_base_test.py [hunk 17]
```diff
@@ -37,7 +37,7 @@ def setUp(self):
         self.temp_dir = tempfile.mkdtemp()
 
     def tearDown(self):
-        mozfile.rmtree(self.temp_dir)
+        mozfile.remove(self.temp_dir)
```
Comment (whimboo) [inline tests/mozhttpd_base_test.py]: Thanks. That will remove those deprecation warnings at least. Once we converted the last unit test over to pytest, we will get rid of this file and the dependency to mozfile.

File: tests/factory/test_factory_invalid_options.py (hunk 9 above)
Comment (rgpt) [reply tests/factory/test_factory_invalid_options.py]: separated them into a separate file you can take look here - https://github.com/rgpt/mozdownload/blob/Convert_directoryParser_factory_toPytest/tests/factory/test_factory_unused_options.py

File: tests/does_not_exist.html (hunk 10 above)
Comment (rgpt) [reply tests/does_not_exist.html]: I was experimenting with the mozdownload's codebase mistakenly added. Will take care of it.

File: tests/mozhttpd_base_test.py (hunk 17 above)
Comment (rgpt) [reply tests/mozhttpd_base_test.py]: I agree and this will increase the code coverage too.

File: tests/factory/test_factory_invalid_options.py (hunk 11 above)
Comment (rgpt) [reply tests/factory/test_factory_invalid_options.py]: renamed.

**Comment by rgpt** [comment]
@whimboo PR updated. Please review.

**Comment by rgpt** [comment]
@whimboo Merging this will complete the porting to pytest process with mozdownload. Please assign me any other issue to work with :)

**Comment by whimboo** [comment]
Sorry for the delay in my response on this PR. I was mostly off last week, and had to catch-up first with a lot of email. I will have a look now.

If you want to work on a different issue on this project please feel free to pick one which you like. Otherwise I'm also happy to show you some others. Best would always be if you could join us on IRC in the #interop channel.

**Comment by whimboo** [comment]
@rgpt please note that when this PR got merged we still have to cleanup the test folder especially ripping out the bast test class, which is no longer needed. Same for the dependency of mozfile.

File: tests/factory/test_factory_unused_options.py (hunk 14 above)
Comment (whimboo) [reply tests/factory/test_factory_unused_options.py]: This is fine for this file. Would you mind to also update it for the others?

File: tests/mozhttpd_base_test.py (hunk 17 above)
Comment (whimboo) [reply tests/mozhttpd_base_test.py]: It shouldn't actually increase the coverage given that the lines stay the same.

File: tests/factory/test_factory_scraper.py [hunk 18]
```diff
@@ -1,44 +1,62 @@
...
 import os
 
-import mozfile
 import pytest
 
 from mozdownload import FactoryScraper
 
 
-@pytest.mark.parametrize('scraper_type,builds,fname,arg', [
+@pytest.mark.parametrize('filename,args', [
```
Comment (whimboo) [inline tests/factory/test_factory_scraper.py]: Why did you change the number of arguments? Note that `args` is used for the arguments as passed to the scraper, but not to push a dict into the test method. You want to revert this specific change, so that it is identical with the other tests.

File: tests/factory/test_factory_unused_options.py [hunk 19]
```diff
@@ -1,72 +1,90 @@
...
 
-import mozfile
 import pytest
 
 from mozdownload.utils import urljoin
 from mozdownload import FactoryScraper
 
 
-@pytest.mark.parametrize('scraper_type,builds,fname,arg', [
+@pytest.mark.parametrize('filename,args', [
```
Comment (whimboo) [inline tests/factory/test_factory_unused_options.py]: Same here in regards of `args`, and the changed number of arguments.

File: tests/factory/test_factory_unused_options.py [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+             'platform': 'mac64',
+             # unused options
+             'build_id': '20131001030204',
+             'build_number': '1',
+             'url': 'http://localhost',
+             'version': '23.0.1',
+         }
+     })
+])
+def test_unknown_options(httpd, tmpdir, filename, args, mocker):
```
Comment (whimboo) [inline tests/factory/test_factory_unused_options.py]: The filename of this test file is `unused`, while here you have `unknown`. Please fix those method names.

File: tests/factory/test_factory_scraper.py (hunk 18 above)
Comment (rgpt) [reply tests/factory/test_factory_scraper.py]: I changed it based on this (https://github.com/mozilla/mozdownload/pull/537#discussion_r258910087) discussion.  Please correct me if i misunderstood something.

File: tests/mozhttpd_base_test.py (hunk 17 above)
Comment (rgpt) [reply tests/mozhttpd_base_test.py]: Sorry my Mistake, Yes the test coverage remains the same because the piece of code does the same thing.

File: tests/factory/test_factory_unused_options.py (hunk 19 above)
Comment (rgpt) [reply tests/factory/test_factory_unused_options.py]: https://github.com/mozilla/mozdownload/pull/537#discussion_r262687689 - So should i keep it same as `filename, args` ?

File: tests/factory/test_factory_invalid_options.py (hunk 9 above)
Comment (rgpt) [reply tests/factory/test_factory_invalid_options.py]: This is incorporated. Please verify and resolve

**Comment by rgpt** [comment]
> Sorry for the delay in my response on this PR. I was mostly off last week, and had to catch-up first with a lot of email. I will have a look now.
> 
> If you want to work on a different issue on this project please feel free to pick one which you like. Otherwise I'm also happy to show you some others. Best would always be if you could join us on IRC in the #interop channel.

No Problem Whatsoever @whimboo. Thank You, I would be very happy to join you on IRC at #interop channel. But i do not know much about "Interop" ?

File: tests/factory/test_factory_unused_options.py (hunk 20 above)
Comment (rgpt) [reply tests/factory/test_factory_unused_options.py]: incorporated this change in the latest commit. Please Review.

File: tests/factory/test_factory_scraper.py (hunk 18 above)
Comment (whimboo) [reply tests/factory/test_factory_scraper.py]: Sorry, I only specified part of the string but not how the whole arguments list should look like. To clarify I meant `scraper_type,builds,filename,args`.

File: tests/factory/test_factory_unused_options.py (hunk 19 above)
Comment (whimboo) [reply tests/factory/test_factory_unused_options.py]: See the other comment.

**Comment by whimboo** [comment]
`#interop` is just the channel name where we are hanging around, and you can reach us for help. It's always helpful to have direct conversations when picking up new issues/bugs. If you need more help in how to join, please let me know.

**Comment by rgpt** [comment]
@whimboo  PR updated. Please review

**Comment by whimboo** [comment]
This looks great. I'm going to merge this PR now.
